sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/estimation/FilterEstimation.scala

```diff
@@ -0,0 +1,479 @@
+        val lit = numericLiteralToBigDecimal(literal, attrRef.dataType)
+        (lit >= statsRange.min) && (lit <= statsRange.max)
+
+      case _ => true  /** for String/Binary type */
+    }
+
+    val percent: Double =
+      if (inBoundary) {
+
+        if (update) {
+          /**
+           * We update ColumnStat structure after apply this equality predicate.
+           * Set distinctCount to 1.  Set nullCount to 0.
+           */
+          val newStats = attrRef.dataType match {
+            case _: NumericType | DateType | TimestampType =>
+              val newValue = Some(literal.value)
+              aColStat.copy(distinctCount = 1, min = newValue,
+                max = newValue, nullCount = 0)
+            case _ => aColStat.copy(distinctCount = 1, nullCount = 0)
+          }
+          mutableColStats += (attrRef.exprId -> newStats)
+        }
+
+        1.0 / ndv.toDouble
+      } else {
+        0.0
+      }
+
+    percent
+  }
+
+  def evaluateInSet(
+      attrRef: AttributeReference,
+      hSet: Set[Any],
+      update: Boolean)
+    : Double = {
+    if (!mutableColStats.contains(attrRef.exprId)) {
+      logDebug("[CBO] No statistics for " + attrRef)
+      return 1.0
+    }
+
+    val aColStat = mutableColStats(attrRef.exprId)
+    val ndv = aColStat.distinctCount
+    val aType = attrRef.dataType
+
+    // use [min, max] to filter the original hSet
+    val validQuerySet = aType match {
+      case _: NumericType | DateType | TimestampType =>
+        val statsRange =
+          Range(aColStat.min, aColStat.max, aType).asInstanceOf[NumericRange]
+        hSet.map(e => numericLiteralToBigDecimal(e, aType, true)).
+          filter(e => e >= statsRange.min && e <= statsRange.max)
+
+      /** We assume the whole set since there is no min/max information for String/Binary type */
+      case StringType | BinaryType => hSet
+    }
+    if (validQuerySet.isEmpty) {
+      return 0.0
+    }
+
+    val newNdv = validQuerySet.size
+    val(newMax, newMin) = aType match {
+      case _: NumericType | DateType | TimestampType =>
+        val tmpSet: Set[Double] = validQuerySet.map(e => e.toString.toDouble)
+        (Some(tmpSet.max), Some(tmpSet.min))
+      case _ =>
+        (None, None)
+    }
+
+    if (update) {
+      val newStats = attrRef.dataType match {
+        case _: NumericType | DateType | TimestampType =>
+          aColStat.copy(distinctCount = newNdv, min = newMin,
+            max = newMax, nullCount = 0)
+        case StringType | BinaryType =>
+          aColStat.copy(distinctCount = newNdv, nullCount = 0)
+      }
+      mutableColStats += (attrRef.exprId -> newStats)
+    }
+
+    /**
+     * return the filter selectivity.  Without advanced statistics such as histograms,
+     * we have to assume uniform distribution.
+     */
+    math.min(1.0, validQuerySet.size / ndv.toDouble)
+  }
+
+  def evaluateBinaryForNumeric(

```

Comment:
Yes, the return value is a double value showing the percentage of rows meeting a given condition. Also I will add comments for this method in JavaDoc style.